File: shared/packets.py
from shared.opcodes import PACKET_STOP
import struct


def packDataRaw(packetFormat: str, data: list):
    headerSize = len(packetFormat) + 1
    headerSize += len(str(headerSize))

    # fix for sizes 11, 101, 102, 1001, 1002, 1003, etc.
    while headerSize < len(f'{headerSize}s' + packetFormat):
        headerSize += 1

    packetFormat = f'{headerSize}s' + packetFormat

    header = bytearray(packetFormat, 'utf-8')
    packet = struct.pack(packetFormat, header, *data)

    return packet + PACKET_STOP


def packData(data: list):
    packetFormat = ''

    for i, d in enumerate(data):
        t = type(d)

        if t == float:
            packetFormat += 'f'
        elif t == int:
            if d < 256 and d >= 0:
                packetFormat += 'B'
            elif d < 127 and d >= -128:
                packetFormat += 'b'
            elif d < 65536 and d >= 0:
                packetFormat += 'H'
            elif d < 32768 and d >= -32768:
                packetFormat += 'h'
            else:
                packetFormat += 'i'
        elif t == bool:
            packetFormat += 'h'
        else:
            if t == str:
                data[i] = data[i].encode('utf-8')

            packetFormat += f'{len(d)}s'

    return packDataRaw(packetFormat, data)


def unpackDataRaw(packet) -> tuple:
    headerSize = int(packet[0:packet.find(b's')])
    header = packet[0:headerSize].decode('utf-8')
    data = struct.unpack(header, packet[:-len(PACKET_STOP)])

    # print('Opcode', opcode)
    # print('Header Size', headerSize)
    # print('Header', header)
    # print('Data', data)

    return data


def unpackData(packet: bytes) -> list:
    data = unpackDataRaw(packet)

    if len(data) < 6:
        raise Exception('Invalid packet, too few arguments')

    if not isinstance(data[1], int):
        raise Exception('Invalid packet, opcode is not an int')
    if not isinstance(data[2], int):
        raise Exception('Invalid packet, origin is not an int')
    if not isinstance(data[3], int):
        raise Exception('Invalid packet, destination is not an int')
    if not isinstance(data[4], int):
        raise Exception('Invalid packet, uniq is not an int')
    if not isinstance(data[5], int):
        raise Exception('Invalid packet, channel is not an int')

    return [data[1], data[2], data[3], data[4], data[5], list(data[6:])]

File: shared-ts/src/packets.ts
import { Opcodes, PACKET_STOP } from './opcodes';
import struct, { DataType } from 'python-struct';
import { Routes } from './routes';
import { Packet } from './interfaces';


export const packDataRaw = (packetFormat: string, ...data: DataType[]): Buffer => {
  let headerSize = packetFormat.length + 1;
  headerSize += headerSize.toString().length;

  // fix for sizes 11, 101, 102, 1001, 1002, 1003, etc.
  while (headerSize < (`${headerSize}s` + packetFormat).length) headerSize += 1;

  const finalFormat = `${headerSize}s` + packetFormat;
  const packet = struct.pack(finalFormat, finalFormat, ...data as DataType[]);


  return Buffer.concat([packet, PACKET_STOP]);
};


type PackData = (opcode: Opcodes, origin: Routes, destination: Routes, uniq: number, channel: Routes, ...data: DataType[]) => Buffer;

export const packData: PackData = (...args): Buffer => {
  let packetFormat = '';

  for (const d of args) {
    if (typeof d == 'number') {
      if (d % 1 === 0) {
        if (d < 256 && d >= 0) {
          packetFormat += 'B';
        } else if (d < 127 && d >= -128) {
          packetFormat += 'b';
        } else if (d < 65536 && d >= 0) {
          packetFormat += 'H';
        } else if (d < 32767 && d >= -32768) {
          packetFormat += 'h';
        } else {
          packetFormat += 'i';
        }
      } else {
        packetFormat += 'f';
      }
    } else if (typeof d === 'boolean') {
      packetFormat += 'h';
    } else {
      packetFormat += `${`${d}`.length}s`;
    }
  }

  return packDataRaw(packetFormat, ...args);
};

export const unpackDataRaw = (packet: Buffer): DataType[] => {
  const headerSize = parseInt(packet.subarray(0, packet.indexOf('s')).toString());
  const header = packet.subarray(0, headerSize).toString();
  const data = struct.unpack(header, packet.subarray(0, -PACKET_STOP.length));

  // console.log(opcode);
  // console.log(headerSize);
  // console.log(header);
  // console.log(data);


  return data;
};

export const unpackData = (packet: Buffer): Packet => {
  const data = unpackDataRaw(packet);

  if (data.length < 6) throw new Error('Invalid packet, too few arguments');

  if (typeof data[1] !== 'number') throw new Error('Invalid packet, opcode is not a number');
  if (typeof data[2] !== 'number') throw new Error('Invalid packet, origin is not a number');
  if (typeof data[3] !== 'number') throw new Error('Invalid packet, destination is not a number');
  if (typeof data[4] !== 'number') throw new Error('Invalid packet, uniq is not a number');
  if (typeof data[5] !== 'number') throw new Error('Invalid packet, channel is not a number');

  const opcode = data[1];
  const origin = data[2];
  const destination = data[3];
  const uniq = data[4];
  const channel = data[5];

  return { opcode, origin, destination, uniq, channel, data: data.slice(6) };
};
